### ISSUE TITLE
[MDK] Example for delaying obfuscation until publish does not work.

### ISSUE THREAD

**Issue by noeppi-noeppi** [MinecraftForge/MinecraftForge#9605]
**Description of issue:**

The MDK shows the following example on how to delay the reobfuscation of the jar file (by the `reobfJar`) task:
```gradle
tasks.named('publish').configure {
    dependsOn 'reobfJar'
}
```
The `maven-publish` plugin will create a separate task (named `publishMavenJavaPublicationToMavenRepository` by default) for each publication. This task actually published the artifact. The `publish` task itself does nothing, it's purpose is to have all those publishing tasks as dependnecies.

Using the shown example from the MDK, both the `publishMavenJavaPublicationToMavenRepository`and the `reobfJar` task will become dependencies of the `publish` task. When executing `publish`, it can happen, that `publishMavenJavaPublicationToMavenRepository` runs *before* `reobfJar`, publishing a deobfuscated jar file to the maven repository.

The correct way to specify that the the `reobfJar` task should run before the publication, is to use the `builtBy` method of `PublicationArtifact` like this:
```gradle
publishing {
    publications {
        mavenJava(MavenPublication) {
            // All the other configuration of the publication
            artifact(jar) {
                builtBy project.provider { project.tasks.named('reobfJar') }
            }
        }
    }
}
```
I think, the MDK should direct the modder to this approach or at least not show an example that does not work and publishes deobfuscated jars into the maven.

**Comment by CombatReq** [comment]
![image](https://github.com/MinecraftForge/MinecraftForge/assets/149759574/1a1a1872-8a62-46d1-ba7f-509fd6e5d0ef)

**Comment by Jonathing** [comment]
Your example works, but does `jar.finalizedBy('reobfJar')` not already do this?

**Comment by noeppi-noeppi** [comment]
Yes, `jar.finalizedBy('reobfJar')` solves this. However, the 1.19 MDK explicitly states that if you want to delay the reobfuscation until publication you can use
```java
tasks.named('publish').configure {
    dependsOn 'reobfJar'
}
```
*instead* of `jar.finalizedBy('reobfJar')`. And replacing `jar.finalizedBy('reobfJar')` with this code snippet breaks. The problem is not the code that is used by default but the commented code that is presented as an alternative when it isn't one.

I see no problem in simply removing the comment that gives modders wrong advice. However, in my opinion it would be a better solution to fix it and a provide a solution that actually works.

**Comment by LexManos** [comment]
Honestly your best bet if you wanna do things weird is to disable reobf and reobf your jar manually.
Reobf does [RenameJarInPlace](https://github.com/MinecraftForge/ForgeGradle/blob/FG_6.0/src/userdev/java/net/minecraftforge/gradle/userdev/tasks/RenameJarInPlace.java#L28) Where you can just make your own [RenameJar](https://github.com/MinecraftForge/ForgeGradle/blob/FG_6.0/src/userdev/java/net/minecraftforge/gradle/userdev/tasks/RenameJar.java#L33 )task

**Comment by Jonathing** [comment]
Removing the comment will be better yeah. 99 out of 100 times, having a deobf compiled jar is a terrible idea, since it would need developers who use it to use whatever mappings it was compiled with. That is what `fg.deobf` is for. Source jars are obviously a different story, depending on the context.

This doesn't apply to 1.20.6+ obviously, but I'll update the MDK sometime later for the older versions to account for this.
